chips/stm32f303xc/src/adc.rs

```diff
@@ -0,0 +1,740 @@
+    Channel6 = 0b00110,
+    Channel7 = 0b00111,
+    Channel8 = 0b01000,
+    Channel9 = 0b01001,
+    Channel10 = 0b01010,
+    Channel11 = 0b01011,
+    Channel12 = 0b01100,
+    Channel13 = 0b01101,
+    Channel14 = 0b01110,
+    Channel15 = 0b01111,
+    Channel16 = 0b10000,
+    Channel17 = 0b10001,
+    Channel18 = 0b10010,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum DiscontinuousMode {
+    OneChannels = 0b000,
+    TwoChannels = 0b001,
+    ThreeChannels = 0b010,
+    FourChannels = 0b011,
+    FiveChannels = 0b100,
+    SixChannels = 0b101,
+    SevenChannels = 0b110,
+    EightChannels = 0b111,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum ExternalTriggerDetection {
+    Disabled = 0b00,
+    RisingEdge = 0b01,
+    FallingEdge = 0b10,
+    RisingAndFalling = 0b11,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum ExternalTriggerSelection {
+    Event0 = 0b0000,
+    Event1 = 0b0001,
+    Event2 = 0b0010,
+    Event3 = 0b0011,
+    Event4 = 0b0100,
+    Event5 = 0b0101,
+    Event6 = 0b0110,
+    Event7 = 0b0111,
+    Event8 = 0b1000,
+    Event9 = 0b1001,
+    Event10 = 0b1010,
+    Event11 = 0b1011,
+    Event12 = 0b1100,
+    Event13 = 0b1101,
+    Event14 = 0b1110,
+    Event15 = 0b1111,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum DataResolution {
+    Bit12 = 0b00,
+    Bit10 = 0b01,
+    Bit8 = 0b10,
+    Bit6 = 0b11,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum ADCStatus {
+    Idle,
+    Off,
+    PoweringOn,
+    OneSample,
+}
+
+pub struct Adc {
+    registers: StaticRef<AdcRegisters>,
+    common_registers: StaticRef<AdcCommonRegisters>,
+    clock: AdcClock,
+    status: Cell<ADCStatus>,
+    client: OptionalCell<&'static dyn EverythingClient>,
+    // clock34: AdcClock,
+}
+
+pub static mut ADC1: Adc = Adc::new();
+
+impl Adc {
+    const fn new() -> Adc {
+        Adc {
+            registers: ADC1_BASE,
+            //rcc copy
+            common_registers: ADC12_COMMON_BASE,
+            clock: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC1)),
+            // clock34: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC34)),
+            status: Cell::new(ADCStatus::Off),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn enable(&self) {
+        self.status.set(ADCStatus::PoweringOn);
+
+        // Enable adc clock
+        self.enable_clock();
+
+        //Set Synchronous clock mode
+        self.common_registers.ccr.modify(CCR::CKMODE.val(0b01));
+
+        self.registers.cr.modify(CR::ADVREGEN.val(0b00));
+        self.registers.cr.modify(CR::ADVREGEN.val(0b01));
+
+        // Wait for ADVRGEN to enable
+        // This needs to be synchronous because there is no interrupt signaling
+        // when ADVRGEN becomes enabled
+        for _i in 0..1000000 {
+            unsafe {
+                llvm_asm!(
+                "nop"
+            : : : : "volatile" );
+            }
+        }
+
+        // Enable the temperature sensor
+        self.common_registers.ccr.modify(CCR::TSEN::SET);
```

Comment:
What does this have to do with the ADC? Is it used invisibly by the hardware to correct for temperature affects somehow?

Reply:
It will be enabled on temperature capsule, it was just for testing. I will do this as soon as I get ADC mux working

Reply:
Should this not be enabled in the temperature capsule then?